File: GameServer/GameServerApp/GameServerApp/Proto/CreateProto/GameServerPageResponseProto.cs
//===================================================
//作    者：xxx
//创建时间：2019-05-18 15:13:23
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary> 请求服务器组列表(返回) </summary>
public struct GameServerPageResponseProto : IProto
{
    public ushort ProtoCode { get { return 50111; } }

    /// <summary> 服务器组数量 </summary>
    public int ItemCount;
    /// <summary> 单组服务器信息 </summary>
    public List<GameServerPageItem> ServerPageItemList;

    /// <summary> 单组服务器信息 </summary>
    public struct GameServerPageItem
    {
        /// <summary> 单组下标 </summary>
        public int PageIndex;
        /// <summary> 单组名称 </summary>
        public string PageServerName;
    }

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(ItemCount);
            for (int i = 0; i < ItemCount; i++)
            {
                ms.WriteInt(ServerPageItemList[i].PageIndex);
                ms.WriteUTF8String(ServerPageItemList[i].PageServerName);
            }
            return ms.ToArray();
        }
    }

    public static GameServerPageResponseProto GetProto(byte[] buffer)
    {
        GameServerPageResponseProto proto = new GameServerPageResponseProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.ItemCount = ms.ReadInt();
            proto.ServerPageItemList = new List<GameServerPageItem>();
            for (int i = 0; i < proto.ItemCount; i++)
            {
                GameServerPageItem _ServerPageItem = new GameServerPageItem();
                _ServerPageItem.PageIndex = ms.ReadInt();
                _ServerPageItem.PageServerName = ms.ReadUTF8String();
                proto.ServerPageItemList.Add(_ServerPageItem);
            }
        }
        return proto;
    }
}

File: MMORPG_2/Assets/Script/Data/Proto/CreateLuaProto/GameServerPageResponseProto.lua
--请求服务器组列表(返回)
GameServerPageResponseProto = { ProtoCode = 50111, ItemCount = 0, ServerPageItemTable = { } }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
GameServerPageResponseProto.__index = GameServerPageResponseProto;

function GameServerPageResponseProto.New()
    local self = { }; --初始化self
    setmetatable(self, GameServerPageResponseProto); --将self的元表设定为Class
    return self;
end


--定义单组服务器信息
ServerPageItem = { PageIndex = 0, PageServerName = "" }
ServerPageItem.__index = ServerPageItem;
function ServerPageItem.New()
    local self = { };
    setmetatable(self, ServerPageItem);
    return self;
end


--发送协议
function GameServerPageResponseProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.ItemCount);
    for i = 1, proto.ItemCount, 1 do
        ms:WriteInt(ServerPageItemList[i].PageIndex);
        ms:WriteUTF8String(ServerPageItemList[i].PageServerName);
    end

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function GameServerPageResponseProto.GetProto(buffer)

    local proto = GameServerPageResponseProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.ItemCount = ms:ReadInt();
	proto.ServerPageItemTable = {};
    for i = 1, proto.ItemCount, 1 do
        local _ServerPageItem = ServerPageItem.New();
        _ServerPageItem.PageIndex = ms:ReadInt();
        _ServerPageItem.PageServerName = ms:ReadUTF8String();
        proto.ServerPageItemTable[#proto.ServerPageItemTable+1] = _ServerPageItem;
    end

    ms:Dispose();
    return proto;
end